Minor TextBoxPainter optimizations

smfr opened the pull request:
#### 8fa2c44953aea49cda4a24ed83b4213f24e351dd
<pre>
Minor TextBoxPainter optimizations
<a href="https://bugs.webkit.org/show_bug.cgi?id=286869">https://bugs.webkit.org/show_bug.cgi?id=286869</a>
<a href="https://rdar.apple.com/144025111">rdar://144025111</a>

Reviewed by Tim Nguyen.

Avoid fetching `m_renderer.frame().editor()` twice in the TextBoxPainter constructor, and in
`paintCompositionForeground()` when we don&apos;t need to paint composition highlights.

* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::m_emphasisMarkExistsAndIsAbove):
(WebCore::TextBoxPainter::paintForegroundAndDecorations):
(WebCore::TextBoxPainter::~TextBoxPainter): Deleted.
* Source/WebCore/rendering/TextBoxPainter.h:

Canonical link: <a href="https://commits.webkit.org/289745@main">https://commits.webkit.org/289745@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99ce817f32e1c59e1c6783eadff5728b645134eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87749 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7265 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42247 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92614 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38499 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89800 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7646 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15436 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67749 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25495 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90751 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5836 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79496 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48118 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5623 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33906 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37606 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76021 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34783 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94500 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14917 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11050 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76599 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15172 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75352 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75835 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18670 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20200 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/18734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7899 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14933 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20236 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14677 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18121 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16459 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -311,7 +311,10 @@ void TextBoxPainter::paintForegroundAndDecorations()
+        if (m_containsComposition)
+            paintCompositionForeground(styledMarkedText);
+        else
+            paintForeground(styledMarkedText);
```
`paintCompositionForeground` already early returns calling `paintForeground` in that case.

If `paintCompositionForeground` isn't to be meant to be called with `!m_containsComposition`, then we should replace the early return with an assert.

If you're trying to avoid a call to `m_renderer.frame().editor()`, then we should just add a fast path in `paintCompositionForeground` when `m_containsComposition` is false.

There are 3 calls to `paintCompositionForeground` so it would be nice to be consistent.

nt1m commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -79,17 +79,17 @@ TextBoxPainter::TextBoxPainter(const LayoutIntegration::InlineContent& inlineCon
 
-TextBoxPainter::~TextBoxPainter()
-{
+    auto& editor = m_renderer.frame().editor();
+    m_containsComposition = m_renderer.textNode() && editor.compositionNode() == m_renderer.textNode();
+    m_useCustomUnderlines = m_containsComposition && editor.compositionUsesCustomUnderlines();
```
I doubt this saves much work tbh, the most expensive bit of LocalFrame::editor() is `ensureEditor()`, which should only be invoked on the first member of this class.

smfr replied:
The cost is jumping from RenderObject to Document to Frame twice.

smfr committed:
Minor TextBoxPainter optimizations
https://bugs.webkit.org/show_bug.cgi?id=286869
rdar://144025111

Reviewed by Tim Nguyen.

Avoid fetching `m_renderer.frame().editor()` twice in the TextBoxPainter constructor, and in
`paintCompositionForeground()` when we don't need to paint composition highlights.

* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::m_emphasisMarkExistsAndIsAbove):
(WebCore::TextBoxPainter::paintForegroundAndDecorations):
(WebCore::TextBoxPainter::~TextBoxPainter): Deleted.
* Source/WebCore/rendering/TextBoxPainter.h:

Canonical link: https://commits.webkit.org/289745@main
